set.mm

```diff
@@ -717097,6 +717100,491 @@ Graph theory (extension)
 =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 $)
 
+$(
+-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-
+  Isomorphic graphs
+-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-
+$)
+
+  $( Introduce new constant symbols. $)
+  $c IsomGr $. $( Predicate "is isomorphic" for graphs $)
+
+  $( Extend class notation to include the "is isomorphic" relation for
+     graphs. $)
+  cisomgr $a class IsomGr $.
+
+  ${
+    $d f g i x y $.
+    $( Define the "is isomorphic" relation for graphs.  (Contributed by AV,
+       11-Nov-2022.) $)
+    df-isomgr $a |- IsomGr = { <. x , y >. |
+                          E. f ( f : ( Vtx ` x ) -1-1-onto-> ( Vtx ` y )
+            /\ E. g ( g : dom ( iEdg ` x ) -1-1-onto-> dom ( iEdg ` y )
+                      /\ A. i e. dom ( iEdg ` x ) ( f " ( ( iEdg ` x ) ` i ) )
+                                        = ( ( iEdg ` y ) ` ( g ` i ) ) ) ) } $.
```

Comment:
A general comment: more important than graph isomorphy is the notion of graph isomorphism.  Indeed, in one direction you can define graph isomorphy as the existence of a graph isomorphism between the two graphs, but the extra information of knowing **how** two things are isomorphic rather than merely that they **are**, can be useful. This is a very general pattern.

Reply:
So you propose to define graph isomorphisms as

```
df-grisom $a |- GrIsom = ( x e. _V , y e. _V |-> { <. f , g >. | ( f : ( Vtx ` x ) -1-1-onto-> ( Vtx ` y )
            /\ ( g : dom ( iEdg ` x ) -1-1-onto-> dom ( iEdg ` y )
                  /\ A. i e. dom ( iEdg ` x ) ( f " ( ( iEdg ` x ) ` i ) ) = ( ( iEdg ` y ) ` ( g ` i ) ) ) ) } )
```

Then `A IsomGr B <-> ( A GrIsom B ) =/= (/)` (By the way, is there at most one graph isomorphism between two graphs?).
I can add this alternate definition/concept to the header comment of the section, and it can be considered if/before this section is moved to main.

Reply:
I haven't checked the precise definition, since I do not know how the various kinds of graphs are encoded in set.mm, but that's the idea. For graph isomorphy, I would use a definition like `A IsomGr B <-> E. f f e. ( A GrIsom B )`, that is, inhabited set versus nonempty set, to be compatible with intuitionistic logic (for when @jkingdon reaches that part of the database!), and as a theorem the version with "nonempty".

Actually, graph isomorphy will be a secondary notion, present only in "final" theorems, but probably most intermediate theorems will use explicitly in their hypotheses or conclusion things like `F e. ( A GrIsom B)`.

Best would probably be to define a category of graphs (there are multiple ones, I haven't looked at the details), where graph morphisms are couples consisting in a function on vertices and a function on edges with required compatibilities, like you wrote. And then, a graph isomorphism is defined as an isomorphism in the category of graphs (something like ``$a |- GraphIsom = ( Isom ` GraphCat ) $.``). I don't remember how much category theory is present in set.mm to make this approach doable/useful, maybe @digama0 knows. With that, you'll be able to use general category theory theorems (provided they are in set.mm). E.g., you'll have that graph isomorphy is an equivalence relation for free.

To your last question: yes, a graph can have multiple automorphisms (take a complete graph and consider vertex permutations).

Reply:
I put most of this material into the section header comment, and also provided a definition for `GrIsom` as described above. This should be sufficient for the moment.